---
title: "Themes, Saving and Interactive Plots"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/Plotting_In_R/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false  
---

<p>&nbsp;</p>
These exercise cover the scales, statistics and themes of ggplot2 for [Plotting in R](https://rockefelleruniversity.github.io/Plotting_In_R/). 


```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```

**Exercise 1** - **Themes**

- Read in the cleaned patients dataset as we saw in ggplot2 course earlier ("patients_clean_ggplot2.txt")

```{r,echo=toMessage}
patients_clean <- read.delim("data/patients_clean_ggplot2.txt",sep="\t")
```

-  Set the global theme to use theme_bw(). Using the patient data set generate a scatter plot of BMI versus Weight. Add a color scale to the scatter plot based on the Pet variable. Use an additional geom to add an extra layer of a fit line to our scatterplot (use lm method). Lets also add a nice color palette of your choice from Colorbrewer, Paleteer or Viridis. 

```{r,echo=toMessage}

library(ggplot2)

theme_set(theme_bw())

plot <- ggplot(data=patients_clean,
               mapping=aes(x=BMI,y=Weight,colour=factor(Pet)))+
  geom_point()+
  stat_smooth(method="lm",se=F) + scale_color_viridis_d(option="plasma")
plot

```

- Remove the legend title from the plot, change the background colors of legend panels to gray and place legend at bottom of plot.

```{r,echo=toMessage}

plot <- plot + theme(legend.title=element_blank(),
                   legend.background=element_rect(fill="gray"),
                   legend.key=element_rect(fill="gray"),
                   legend.position="bottom")

plot
```

- Remove the legend title from the plot, change the background colors of legend panels to white and place legend at bottom of plot.

```{r,echo=toMessage}

plot <- plot+facet_wrap(~Pet)

plot
```

- Add a title to the plot, remove minor grid lines and save to 7 by 7 inch png file.

```{r,echo=toMessage}

plot <- plot+ggtitle("BMI vs Weight - By Pet")+theme(panel.grid.minor=element_blank())

plot

ggsave(plot,file="BMIvsWeight.png",units = "in",height = 7,width = 7)
```

- Save our current theme then try out some alternative themes. Try one theme included with ggplot2 and one from ggthemes.

```{r,echo=toMessage}
myTheme <- plot$theme

plot + theme_linedraw()


library(ggthemes)
plot + theme_fivethirtyeight()

```

- Take at least one element from each theme you just tried out and add it to our existing theme.

```{r,echo=toMessage}

myTheme <- myTheme + theme(strip.background = element_rect(fill = "black"), # from theme_linedraw
            strip.text = element_text(colour = "white", size = rel(0.8)), # from theme_linedraw
             panel.grid = element_line(colour = NULL), # from theme_fivethirtyeight
            panel.grid.major = element_line(colour = "#D2D2D2"), # from theme_fivethirtyeight
            panel.grid.minor = element_blank(),# from theme_fivethirtyeight
            rect = element_rect(fill = "#F0F0F0", # from theme_fivethirtyeight
            linetype = 0, colour = NA))# from theme_fivethirtyeight

```

- Use the + to update our plot to use the new theme

```{r,echo=toMessage}

plot + myTheme

```

- Save the final plot as a pdf. 

```{r,echo=toMessage}
ggsave(plot, file="BMIvsWeight.pdf")


```


- Reset the theme to the default: theme_gray()

```{r,echo=toMessage}
theme_set(theme_gray())

plot

```

**Exercise 2** - **External Packages**

- Load in the data from "data/DESeq2_PCA.RData". This contains a ggplot object that is produced by the package DESeq2. It is a PCA plot produced from differential gene expression analysis. Review the object. Check the plot, data, aesthetics etc. 

```{r, eval=F, echo=F, include =F}
dds <- makeExampleDESeqDataSet(betaSD=1)
rld <- rlog(dds)
plotPCA(rld)

# also possible to perform custom transformation:
dds <- estimateSizeFactors(dds)
# shifted log of normalized counts
se <- SummarizedExperiment(log2(counts(dds, normalized=TRUE) + 1),
                           colData=colData(dds))
# the call to DESeqTransform() is needed to
# trigger our plotPCA method.
test <- plotPCA( DESeqTransform( se ) )
test$data$condition <- rep(c(rep("Ctrl",3),rep("Mut",3)),2)
deseq_pca_example <- test
save(deseq_pca_example, file ="data/DESeq2_PCA.RData")

dds <- DESeq(dds)

library(org.Mm.eg.db)



my_res <- results(dds, contrast = c("condition", "A","B"))

rownames(my_res) <- sample(keys(org.Mm.eg.db, keytype = "SYMBOL"),1000)

my_res <- as.data.frame(my_res)

my_res$pvalue[is.na(my_res$pvalue)] <- rep(0.5,5)

my_res$SYMBOL <- rownames(my_res)

save(my_res,file="../extdata/data/DESeq2_Result.RData")
```

```{r,echo=toMessage}
load("data/DESeq2_PCA.RData")

deseq_pca_example

deseq_pca_example$data

deseq_pca_example$mapping
```

- Lets change the color scheme and add a title. 

```{r,echo=toMessage}

deseq_pca_example + scale_fill_viridis_d() + ggtitle("PCA plot of DGE in A and B groups")

```

- When we look at the data column we can see there is an additional column contained in the metadata - condition. Lets add this to the plot. 

```{r,echo=toMessage}
deseq_replot <- deseq_pca_example
deseq_replot$layers <- NULL

deseq_replot <- deseq_replot + geom_point(aes(x=PC1, y=PC2, color=group, shape=condition)) + scale_fill_viridis_d() + ggtitle("PCA plot of DGE in A and B groups")

```

**Exercise 3** - **Interactive Plots**

- Turn our PCA plot into an interactive plot. 

```{r,echo=toMessage}
library(plotly)

ggplotly(deseq_replot)
```


- Add the names of each sample from the orginal data object to the interactive labeling. 

```{r,echo=toMessage}
ggplotly(deseq_replot + geom_point(aes(label = name)))


```

**Exercise 4** - **Working Example**

In this final exercise we will run through a common example: making a volcano plot.

A volcano plot consists of:
1) log2FC on the x axis
2) -log10(pval) on the y axis

It is also good to add some additional customization:
- Highlight significant genes i.e. pval <0.05
- Highlight genes above a certain threshold log2FC i.e. >1
- Add lines to denote these thresholds
- Label some genes of interest directly: "Gm8714","Pas1b","Rab39","Tmc2","Ttpal","Ctdsp1"
- Use a simple theme to also give the plot a simple look
- Export the plot as a pdf

```{r,echo=toMessage}

load("data/DESeq2_Result.RData")
```


```{r,echo=toMessage}
library(plotly)

goi <- c("Gm8714","Pas1b","Rab39","Tmc2","Ttpal","Ctdsp1")

myplot <- ggplot(my_res, aes(x=log2FoldChange, 
                   y=-log10(pvalue), 
                   color = ifelse(pvalue > 0.05, "NS",
                                 ifelse(log2FoldChange > 1, "SigUp", 
                                        ifelse(log2FoldChange < (-1), "SigDown", "Sig"))))) +
  geom_point(size=0.5, alpha=0.5) +
  scale_color_manual(name = "Significance", breaks=c("NS", "Sig", "SigUp", "SigDown"), values = c("black", "green","blue", "red")) +
  theme_bw() + 
  ggtitle("Volcano Plot showing significance of \ngene expression changes following DESeq analysis") + 
  geom_hline(yintercept=(-log10(0.05)), lty =3, color="gray") + 
  geom_vline(xintercept=c((-1),1), lty =3, color="gray") + 
  geom_text(aes(label=ifelse(SYMBOL %in% goi, as.character(SYMBOL),''))) + xlim(-4.5,4.5)

myplot

ggsave(myplot, "DESeq2_volcanoplot.pdf")
```

- To improve labeling on ggplot there is an additional package that can be useful: ggrepel. Try installing ggrepel and using geom_text_repel.

```{r,echo=toMessage}
library(ggrepel)

goi <- c("Gm8714","Pas1b","Rab39","Tmc2","Ttpal","Ctdsp1")

ggplot(my_res, aes(x=log2FoldChange, 
                   y=-log10(pvalue), 
                   color = ifelse(pvalue > 0.05, "NS",
                                 ifelse(log2FoldChange > 1, "SigUp", 
                                        ifelse(log2FoldChange < (-1), "SigDown", "Sig"))))) +
  geom_point(size=0.5, alpha=0.5) +
  scale_color_manual(name = "Significance", breaks=c("NS", "Sig", "SigUp", "SigDown"), values = c("black", "green","blue", "red")) +
  theme_bw() + 
  ggtitle("Volcano Plot showing significance of \ngene expression changes following DESeq analysis") + 
  geom_hline(yintercept=(-log10(0.05)), lty =3, color="gray") + 
  geom_vline(xintercept=c((-1),1), lty =3, color="gray") + 
  geom_text_repel(aes(label=ifelse(SYMBOL %in% goi, as.character(SYMBOL),'')),  min.segment.length = 0, seed = 42, box.padding = 0.2, max.time	=3,max.overlaps = Inf) + xlim(-4.5,4.5)

```

- Make it dynamic and interactive with plotly just labeling the SYMBOL information.

```{r}


myplot <- ggplot(my_res, aes(x=log2FoldChange, 
                   y=-log10(pvalue), 
                   color = ifelse(pvalue > 0.05, "NS",
                                 ifelse(log2FoldChange > 1, "SigUp", 
                                        ifelse(log2FoldChange < (-1), "SigDown", "Sig"))))) +
  geom_point(size=0.5, alpha=0.5) +
  scale_color_manual(name = "Significance", breaks=c("NS", "Sig", "SigUp", "SigDown"), values = c("black", "green","blue", "red")) +
  theme_bw() + 
  ggtitle("Volcano Plot showing significance of \ngene expression changes following DESeq analysis") + 
  geom_hline(yintercept=(-log10(0.05)), lty =3, color="gray") + 
  geom_vline(xintercept=c((-1),1), lty =3, color="gray") + 
  theme(text = element_text(size = 8))

ggplotly(myplot + geom_point(aes(text = SYMBOL )), source = "select", tooltip = c("SYMBOL"))
```


