<!DOCTYPE html>
<html>
  <head>
    <title>Introduction to Plotting</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Introduction to Plotting
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Plotting_In_R/" class="uri">http://rockefelleruniversity.github.io/Plotting_In_R/</a>

---

class: inverse, center, middle

# Plotting in R (Base graphics)
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---
##Introduction




R has excellent graphics and plotting capabilities. They are mostly found in following three sources.
 + base graphics
 + the lattice package
 +  the ggplot2 package
 
Base R graphics uses a pen and paper model for plotting while Lattice and ggplot2 packages are built on the routines first used in grid graphics.

---
## A pen and paper model

- Once plot is produced, can only add more elements, cannot remove.
- Makes it hard to update.
- But faster than more complex plotting systems.



---

## Scatter and linne Charts

First we'll produce a very simple graph using the values in a numeric vector:


```r
treatment &lt;- c(0.02,1.8, 17.5, 55,75.7, 80)
```

---
##Scatter Plot

Now we plot the treatment vector with default parameters


```r
plot(treatment)
```



&lt;img src="basePlotting_files/figure-html/unnamed-chunk-3-1.png" width="1920px" /&gt;

---
Now, let's customise this plot a little. 

First we can plot treatment using points overlayed by a line. We control this with the **type** argument.



```r
plot(treatment, type="o")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-4-1.png" width="1000px" /&gt;

---
We have many type options available to us. 

To see a complete list we can use **?plot**

.pull-left[

```r
plot(treatment, type="l")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-5-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, type="p")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-6-1.png" width="1000px" /&gt;
]
---

We can add additional labels to our plot's axis and main/sub-title too.

We add a title with **main** argument and or a sub-title with the **sub** argument.


```r
plot(treatment, main="My Plot", sub="a plot")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-7-1.png" width="1000px" /&gt;

---

We can customise our x and y axis label with the **xlab** and **ylab** arguments respectively.


```r
plot(treatment, xlab="Position", ylab="score")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-8-1.png" width="1000px" /&gt;

---

We can control the orientation of labels on axis using **las** argument.

.pull-left[

```r
plot(treatment, las=1)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-9-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, las=2)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-10-1.png" width="1000px" /&gt;
]

---

We can control the size of points in our plot using the **cex** parameter.

.pull-left[

```r
plot(treatment, cex=2)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-11-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, cex=0.5)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-12-1.png" width="1000px" /&gt;
]

---

We can control the type of points in our plot using the **pch** parameter.

.pull-left[

```r
plot(treatment, pch=1)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-13-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, pch=20)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-14-1.png" width="1000px" /&gt;
]


---

Similarly when plotting a line we control size with **lwd** parameter.

.pull-left[

```r
plot(treatment, type="l",lwd=10)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-15-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, type="l",lwd=0.5)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-16-1.png" width="1000px" /&gt;
]

---

We can also control the type of line with **lty** parameter.

.pull-left[

```r
plot(treatment, type="l",lty=1)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-17-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, type="l",lty=2)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-18-1.png" width="1000px" /&gt;
]

---


An important parameter we can control is colour. We can control colour or lines or points using the **col** argument.

.pull-left[

```r
plot(treatment, type="l",
     col="red")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-19-1.png" width="1000px" /&gt;
]
.pull-right[

```r
plot(treatment, type="l",
     col="dodgerblue")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-20-1.png" width="1000px" /&gt;
]

---

You can find an extensive list of R colours here.

[R colour Chart](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf)

---


Review **?plot** and **?par** for complete list of options.


---

The **plot** function vector will accept two vectors to be plotted against each other.


```r
control &lt;- c(0, 20, 40, 60, 80,100)
plot(treatment,control)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-21-1.png" width="1000px" /&gt;

---

But we often want mulitple lines in same plot. So if we want to plot scores for control and treatment against position we will need a new method.

We can add an addition line to our existing plot using the **lines()** function.


```r
plot(treatment, type="o", col="blue")
lines(control, type="o", pch=22, lty=2, col="red")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-22-1.png" width="1000px" /&gt;

---
But the new line doesnt quite fit into our original plot.


We can extend our x or y axis by specifying values to **xlim** **ylim** arguments.


```r
control &lt;- c(0, 20, 40, 60, 80,100)
plot(treatment, type="o", col="blue",ylim=c(0,100))
lines(control, type="o", pch=22, lty=2, col="red")
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-23-1.png" width="1000px" /&gt;

---

Next let's change the axes labels to match our data and add a legend. We'll also compute the y-axis values using the **range** function 
so any changes to our data will be automatically reflected in our graph. 

range() returns a vector containing the minimum and maximum of all the given arguments.


Calculate range from 0 to max value of data

```r
g_range &lt;- range(0, treatment, control)
g_range
```

```
## [1]   0 100
```


---

Plot treatment using y axis that ranges from 0 to max value in treatment or control vector.  Turn off axes and annotations (axis labels) so we can specify them ourselves. We turn of axis and annotation plotting using **axes=FALSE** and **ann=FALSE**


```r
plot(treatment, type="o", col="blue", 
     ylim=g_range,axes=FALSE, ann=FALSE)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-25-1.png" width="1000px" /&gt;

---

We can now create our own X axis by using the **axis()** function. We specify the **side** argument for where to place axis, the **at** argument to specify where to put axis ticks and **lab** argument to specify labels for axis ticks.


```r
axis(side=1, at=1:6, lab=c("Mon","Tue","Wed","Thu","Fri","Sat"))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-27-1.png" width="1000px" /&gt;

---

We can make our y axis with horizontal labels that display ticks at every 20 marks in a similar way.

We specify our **side** and use **rep()** function to make axis tick postions for **at** argument.


```r
axis(2, las=1, at=rep(0,g_range[2],by=20))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-29-1.png" width="1000px" /&gt;

---

We can now add a box around our plot using the **box()** function.

```r
box()
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-31-1.png" width="1000px" /&gt;


---

Now i can add my control data using lines argument.


```r
lines(control, type="o", pch=22, lty=2, col="red")

```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-33-1.png" width="1920px" /&gt;
---

Finally we may wish to add a legend to out plot. We can add a legend to current plot using the **legend()** function.

We need to specify where to place legend in plot, the names in legend to **legend** argument and any additional point/line type configuration we used, e.g blue/red.


```r
legend("topleft",legend=c("treatment","control"),
       col=c("blue","red"), pch=21:22, lty=1:2);  
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-35-1.png" width="1920px" /&gt;
---

##Bar Charts	
	
	
Base graphics has a useful built in function for bar charts. The **barchart()** function.
We can simply pass our numeric vector to this function to get our barchart.


```r
barplot(treatment)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-37-1.png" width="1920px" /&gt;

---

The **barchart()** function hasn't added any labels by default. We can speciy our own however using the **names.arg** argument. **names.arg**  is a vector of names to be plotted below each bar or group of bars. 



```r
barplot(treatment,
        names.arg=c("Mon","Tue","Wed","Thu","Fri","Sat"))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-38-1.png" width="1920px" /&gt;

---

If my vector was named however, then my vectors names would be used for labels. We use **names()** function to add names to our vector then we replot.



```r
names(treatment) &lt;- c("Mon","Tue","Wed","Thu","Fri","Sat")
barplot(treatment)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-39-1.png" width="1920px" /&gt;


---
 
Let's now read the data from the example.txt data file.  

Read values from tab-delimited example.txt

```r
data &lt;- read.table("data/example.txt", header=T, sep="\t")
```



---
We can now plot data from a matrix with a stacked barchart. Note we use **as.matrix()** function to convert our data.frame to a matrix.



```r
barplot(as.matrix(data))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-41-1.png" width="1920px" /&gt;

---
Now we can plot data from a matrix with side-by-side barchart using the **beside** argument



```r
barplot(as.matrix(data),beside=TRUE)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-42-1.png" width="1920px" /&gt;

---
Now we can put it together and plot data with adjacent bars using colors and place the legend at the top-left corner.



```r
barplot(as.matrix(data), main="Data", ylab= "Total", beside=TRUE, 
        col= c("lightblue", "mistyrose", "lightcyan","lavender",
               "cornsilk","maroon"))
legend("topleft", c("Mon","Tue","Wed","Thu","Fri","Sat"), cex=0.8,
        fill=  c("lightblue", "mistyrose", "lightcyan",
                         "lavender","cornsilk","maroon"));
```


&lt;img src="basePlotting_files/figure-html/unnamed-chunk-44-1.png" width="1920px" /&gt;

---

##Histograms


Base graphics has a useful built in function for histograms too. The **histogram()** function.
We can simply pass our numeric vector to this function to get our barchart.


```r
hist(treatment)  
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-45-1.png" width="1920px" /&gt;


---

Similar cutomisation exists as for other plots.

```r
hist(treatment, col="lightblue", ylim=c(0,10))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-47-1.png" width="1920px" /&gt;
---

We can create more fine grained histogram by specify the number of required bins to the **breaks** argument.

```r
hist(treatment, col="lightblue",breaks = 2)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-49-1.png" width="1920px" /&gt;

---

##Dot charts

Base graphics has a useful built in function for dotcharts too. The **dotchart()** function.
We can simply pass our numeric vector to this function to get our barchart.

Here we use the function **t** to return the transpose of a matrix.

```r
dotchart(t(data))	
```
&lt;img src="basePlotting_files/figure-html/unnamed-chunk-51-1.png" width="1920px" /&gt;

---

Let's make the dotchart a little more colourful: 

Now we create a coloured dotchart for autos with smaller labels

```r
dotchart(t(data), color=c("red","blue"),
         main="Dotchart", cex=0.8)	
```
&lt;img src="basePlotting_files/figure-html/unnamed-chunk-53-1.png" width="1920px" /&gt;

---

##Box plots

The final plot we will look at is a box and whisker plot. 

Boxplots allow you to quickly review data distributions, showing the median and 1st/3rd quartile.

![](imgs/box_whisker_describe.png)

---

First lets read in the gene expression data

```r
exprs &lt;- read.delim("data/gene_data.txt",sep="\t",h=T,row.names = 1)
head(exprs)
```

```
##                    Untreated1 Untreated2  Treated1   Treated2
## ENSDARG00000093639  0.8616832  1.9311442 0.1041508 0.14055604
## ENSDARG00000094508  0.9857575  2.0256352 0.1549917 0.20301609
## ENSDARG00000095893  0.8498889  1.9875580 0.2317969 0.20925123
## ENSDARG00000095252  0.9242996  2.0857620 0.2562264 0.24669079
## ENSDARG00000078878  0.3571734  0.4653908 0.1167221 0.09710237
## ENSDARG00000079403  1.0604071  1.2581398 0.3884836 0.31567299
```

---

Now we can use the **boxplot()** function on our data.frame to get our boxplot

```r
boxplot(exprs)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-55-1.png" width="1920px" /&gt;

---

Perhaps it would look better on a log scale. We can add addition colours and labels as with other plots.


```r
boxplot(log2(exprs),ylab="log2 Expression",
        col=c("red","red","blue","blue"))
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-56-1.png" width="1920px" /&gt;

---
##Combining Plots

R makes it easy to combine multiple plots into one overall graph, using either the **par( )** or **layout( )** function. 
With the **par( )** function, you can include the option mfrow=c(nrows, ncols) to create a matrix of nrows x ncols plots that are filled in by row.
mfcol=c(nrows, ncols) fills in the matrix by columns.

Define a layout with 2 rows and 2 columns


```r
par(mfrow=c(2,2))
```

---

Here, we will use different dataset with two columns each for treated and untreated samples.


```r
data1 &lt;- read.table("data/gene_data.txt", header=T, sep="\t")
head(data1)
```

```
##      ensembl_gene_id Untreated1 Untreated2  Treated1   Treated2
## 1 ENSDARG00000093639  0.8616832  1.9311442 0.1041508 0.14055604
## 2 ENSDARG00000094508  0.9857575  2.0256352 0.1549917 0.20301609
## 3 ENSDARG00000095893  0.8498889  1.9875580 0.2317969 0.20925123
## 4 ENSDARG00000095252  0.9242996  2.0857620 0.2562264 0.24669079
## 5 ENSDARG00000078878  0.3571734  0.4653908 0.1167221 0.09710237
## 6 ENSDARG00000079403  1.0604071  1.2581398 0.3884836 0.31567299
```
---

Plot histograms for different columns in the data frame separately. This is not very efficient. 
You could also do it more efficiently using for loop.


```r
par(mfrow=c(2,2))
hist(data1$Untreated1)
hist(data1$Treated2)
hist(data1$Untreated2)
boxplot(data1$Treated1)
```

&lt;img src="basePlotting_files/figure-html/unnamed-chunk-60-1.png" width="490px" /&gt;

---
##Saving your plots



There are many different ways of saving your plots in R. 

The only argument you would need is name of file in which you want to save the plot.

Plotting commands then can be entered as usual.
The output would be redirected to the file. 

When you're done with your plotting commands, enter the dev.off() command. 


```r
bmp(filename, width = 480, height = 480, units = "px", 
    point-size = 12)
jpeg(filename, width = 480, height = 480, units = "px", 
     point-size = 12, quality = 75)
```
---

Saving in bitmap format

```r
bmp(file = "control.bmp")
plot(control)
dev.off()
```
---
Saving in jpeg format

```r
jpeg(file = "control.jpg", quality = 20)
plot(control)
dev.off()
```

---

Saving in postscript format


```r
postscript(file = "control.ps")
plot(control)
dev.off()
```

---
saving in pdf format


```r
pdf(file = "control.pdf", paper = "A4")
plot(control)
dev.off()
```


---

Exercise on base plotting can be found [here](../../exercises/Plotting_exercise.html)

---

Answers for baseplotting can be found [here](../../answers/Plotting_answers.html)

---
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
